src/hid-asus.c

```diff
@@ -0,0 +1,320 @@
+#define CONTACT_Y_MSB_MASK 0x0f
+#define CONTACT_Y_LSB_OFFSET 2
+
+#define CONTACT_TOUCH_MAJOR_OFFSET 3
+#define CONTACT_TOUCH_MAJOR_BIT_SHIFT 4
+#define CONTACT_TOUCH_MAJOR_MASK 0x07
+
+#define CONTACT_PRESSURE_OFFSET 4
+#define CONTACT_PRESSURE_MASK 0x7f
+
+#define BYTE_BIT_SHIFT 8
+#define TRKID_SGN       ((TRKID_MAX + 1) >> 1)
+
+static void asus_report_contact_down(struct input_dev *input,
+		int toolType, u8 *data)
+{
+	int x, y, touch_major, pressure;
+
+	x = data[CONTACT_X_MSB_OFFSET] >> CONTACT_X_MSB_BIT_SHIFT;
+	x = (x << BYTE_BIT_SHIFT) | data[CONTACT_X_LSB_OFFSET];
+
+	y = (data[CONTACT_Y_MSB_OFFSET] & CONTACT_Y_MSB_MASK);
+	y = MAX_Y - ((y << BYTE_BIT_SHIFT) | data[CONTACT_Y_LSB_OFFSET]);
+
+	if (toolType == MT_TOOL_PALM) {
+		touch_major = MAX_TOUCH_MAJOR;
+		pressure = MAX_PRESSURE;
+	} else {
+		touch_major = data[CONTACT_TOUCH_MAJOR_OFFSET];
+		touch_major >>= CONTACT_TOUCH_MAJOR_BIT_SHIFT;
+		touch_major &= CONTACT_TOUCH_MAJOR_MASK;
+		pressure = data[CONTACT_PRESSURE_OFFSET] & CONTACT_PRESSURE_MASK;
+	}
+
+	input_report_abs(input, ABS_MT_POSITION_X, x);
+	input_report_abs(input, ABS_MT_POSITION_Y, y);
+	input_report_abs(input, ABS_MT_TOUCH_MAJOR, touch_major);
+	input_report_abs(input, ABS_MT_PRESSURE, pressure);
+}
+
+/* Required for Synaptics Palm Detection */
+static void asus_report_tool_width(struct input_dev *input)
+{
+	struct input_mt *mt = input->mt;
+	struct input_mt_slot *oldest;
+	int oldid, count, i;
+
+	oldest = NULL;
+	oldid = mt->trkid;
+	count = 0;
+
+	for (i = 0; i < mt->num_slots; ++i) {
+		struct input_mt_slot *ps = &mt->slots[i];
+		int id = input_mt_get_value(ps, ABS_MT_TRACKING_ID);
+
+		if (id < 0)
+			continue;
+		if ((id - oldid) & TRKID_SGN) {
+			oldest = ps;
+			oldid = id;
+		}
+		count++;
+	}
+
+	if (oldest) {
+		input_report_abs(input, ABS_TOOL_WIDTH,
+			input_mt_get_value(oldest, ABS_MT_TOUCH_MAJOR));
+	}
+}
+
+static void asus_report_input(struct input_dev *input, u8 *data)
+{
+	int i;
+	u8 *contactData = data + CONTACT_DATA_OFFSET;
+
+	for (i = 0; i < MAX_CONTACTS; i++) {
+		bool down = data[CONTACT_DOWN_OFFSET] &
+				(CONTACT_DOWN_MASK << i);
+		int toolType = contactData[CONTACT_TOOL_TYPE_OFFSET] &
+				CONTACT_TOOL_TYPE_MASK ? MT_TOOL_PALM
+				: MT_TOOL_FINGER;
+
+		input_mt_slot(input, i);
+		input_mt_report_slot_state(input, toolType, down);
+
+		if (down) {
+			asus_report_contact_down(input, toolType, contactData);
+			contactData += CONTACT_DATA_SIZE;
+		}
+	}
+
+	input_report_key(input, BTN_LEFT,
+			data[BTN_LEFT_OFFSET] & BTN_LEFT_MASK);
+	asus_report_tool_width(input);
+
+	input_mt_sync_frame(input);
+	input_sync(input);
+}
+
+static int asus_raw_event(struct hid_device *hdev,
+		struct hid_report *report, u8 *data, int size)
+{
+	if (hdev->product == USB_DEVICE_ID_ASUSTEK_TOUCHPAD &&
+			 data[REPORT_ID_OFFSET] == INPUT_REPORT_ID &&
+						size == INPUT_REPORT_SIZE) {
+		struct hid_input *hidinput;
+
+		list_for_each_entry(hidinput, &hdev->inputs, list) {
```

Comment:
No need to look for the input. IIRC before you stored it in the driver data, and that's easier and simpler.

Also, to support the Bluetooth keyboard combo from the T300CHI, you'll need to have more than one input node, so it's just easier to store the touch input node directly.